From 0e9169b6e889a4829d29a4e7263f71634e0709a4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Wed, 4 Dec 2024 22:34:56 +0100
Subject: [PATCH 65/86] gdi32: Return immediately with negative count.

gdi32_test font (ASan)

ASAN_OPTIONS='verbosity=0:windows_hook_rtl_allocators=1' WINEDLLOVERRIDES="$F=n;*.dll=n" WINEDEBUG= wine64 z:/home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj/dlls/gdi32/tests/x86_64-windows/gdi32_test.exe font
...
font.c:1213: Tests skipped: TrueType font for charset 136 is not installed
=================================================================
==1536==ERROR: AddressSanitizer: requested allocation size 0xfffffffffffffffe (0x800 after adjustments for alignment, red zones etc.) exceeds maximum supported size of 0x10000000000 (thread T0)
0614:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffffa61a8a3 in HeapAlloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:230:3
    #1 0x6ffffc4c9eff in BIDI_Reorder.isra.0 /home/bernhard/data/entwicklung/2024/wine/wine/dlls/gdi32\text.c:397:16
    #2 0x6ffffc4e5f74 in GetTextExtentExPointW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/gdi32\text.c:1393:5
    #3 0x0001400cc8c1 in test_text_extents /home/bernhard/data/entwicklung/2024/wine\wine/dlls/gdi32/tests/font.c:1481:11
    #4 0x0001400cc8c1 in func_font /home/bernhard/data/entwicklung/2024/wine\wine/dlls/gdi32/tests/font.c:7868:5
    #5 0x0001401bcac1 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #6 0x0001401bcac1 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #7 0x0001401be9cf in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #8 0x6fffffa54808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #9 0x6fffffc2fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

==1536==HINT: if you don't care about these errors you may set allocator_may_return_null=1
SUMMARY: AddressSanitizer: allocation-size-too-big /home/bernhard/data/entwicklung/2024/wine/wine/dlls/gdi32\text.c:397:16 in BIDI_Reorder.isra.0
==1536==ABORTING
---
 dlls/gdi32/text.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/gdi32/text.c b/dlls/gdi32/text.c
index 2de68f892d8..815b282ef4f 100644
--- a/dlls/gdi32/text.c
+++ b/dlls/gdi32/text.c
@@ -1384,6 +1384,7 @@ BOOL WINAPI GetTextExtentExPointW( HDC hdc, const WCHAR *str, INT count, INT max
     BOOL ret = TRUE;
 
     if (!(dc_attr = get_dc_attr( hdc ))) return FALSE;
+    if (count < 0) return FALSE;
 
     bidi_flags = dc_attr->text_align & TA_RTLREADING ? WINE_GCPW_FORCE_RTL : WINE_GCPW_FORCE_LTR;
     bidi_flags |= WINE_GCPW_DISABLE_REORDERING;
-- 
2.47.1

